Refactor GitRepositoryExtension to avoid monkey-patching

dduugg opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----
This refactors `GitRepositoryExtension` as a static module, rather than one that was dynamically extended by a `Pathname`. This approach allows us to enable typing, remove an `.rbi` file, and make the code more comprehensible (it wasn't clear which `Pathname` objects had this add'l behavior).

Note that `Library/Homebrew/extend/git_repository.rb` and `Library/Homebrew/utils/git_repository.rb` have considerable overlap, but i haven't attempted to unify them in this PR. (The `extend` version lacks test, the `utils` version is part of the public API, and they diverge in subtle ways. For instance, both have a `git_commit_message` but with opposite defaults for the `safe` kwarg.)

MikeMcQuaid commented on Library/Homebrew/extend/git_repository.rb:
```diff
@@ -4,111 +4,111 @@
-  def git?
-    join(".git").exist?
+  sig { params(pathname: Pathname).returns(T::Boolean) }
+  def self.git?(pathname)
```
A few name suggestions

```suggestion
  def self.dot_git?(pathname)
```

```suggestion
  def self.git_directory?(pathname)
```

```suggestion
  def self.exists?(pathname)
```

```suggestion
  def self.valid?(pathname)
```

dduugg replied:
I went with `.git_repo?` bc i think it's the best description of what it's verifying. wdyt?

MikeMcQuaid commented on Library/Homebrew/extend/git_repository.rb:
```diff
@@ -4,111 +4,111 @@
-  def git_origin
-    popen_git("config", "--get", "remote.origin.url")
+  sig { params(pathname: Pathname).returns(T.nilable(String)) }
+  def self.git_origin(pathname)
```
```suggestion
  def self.origin(pathname)
```

or 

```suggestion
  def self.origin_url(pathname)
```

MikeMcQuaid commented on Library/Homebrew/extend/git_repository.rb:
```diff
@@ -4,111 +4,111 @@
-  def git_origin=(origin)
-    return if !git? || !Utils::Git.available?
+  sig { params(pathname: Pathname, origin: String).returns(T.nilable(T::Boolean)) }
+  def self.git_origin=(pathname, origin)
```
```suggestion
  def self.origin=(pathname, origin)
```

or 

```suggestion
  def self.origin_url=(pathname, origin_url)
```

MikeMcQuaid commented on Library/Homebrew/extend/git_repository.rb:
```diff
@@ -4,111 +4,111 @@
-  def git_head(safe: false)
-    popen_git("rev-parse", "--verify", "--quiet", "HEAD", safe: safe)
+  sig { params(pathname: Pathname, safe: T::Boolean).returns(T.nilable(String)) }
+  def self.git_head(pathname, safe: false)
```
```suggestion
  def self.head(pathname, safe: false)
```
or
```suggestion
  def self.head_ref(pathname, safe: false)
```

MikeMcQuaid commented on Library/Homebrew/extend/git_repository.rb:
```diff
@@ -4,111 +4,111 @@
-  sig { params(length: T.nilable(Integer), safe: T::Boolean).returns(T.nilable(String)) }
-  def git_short_head(length: nil, safe: false)
+  sig { params(pathname: Pathname, length: T.nilable(Integer), safe: T::Boolean).returns(T.nilable(String)) }
+  def self.git_short_head(pathname, length: nil, safe: false)
```
```suggestion
  def self.short_head(pathname, length: nil, safe: false)
```

or

```suggestion
  def self.short_head_ref(pathname, length: nil, safe: false)
```

MikeMcQuaid commented on Library/Homebrew/extend/git_repository.rb:
```diff
@@ -4,111 +4,111 @@
-  def git_last_commit
-    popen_git("show", "-s", "--format=%cr", "HEAD")
+  sig { params(pathname: Pathname).returns(T.nilable(String)) }
+  def self.git_last_commit(pathname)
```
```suggestion
  def self.last_commit(pathname)
```
or
```suggestion
  def self.show_last_commit(pathname)
```
or
```suggestion
  def self.show(pathname)
```

MikeMcQuaid commented on Library/Homebrew/extend/git_repository.rb:
```diff
@@ -4,111 +4,111 @@
-  def git_branch(safe: false)
-    popen_git("rev-parse", "--abbrev-ref", "HEAD", safe: safe)
+  sig { params(pathname: Pathname, safe: T::Boolean).returns(T.nilable(String)) }
+  def self.git_branch(pathname, safe: false)
```
```suggestion
  def self.branch(pathname, safe: false)
```
or
```suggestion
  def self.branch_name(pathname, safe: false)
```

MikeMcQuaid commented on Library/Homebrew/extend/git_repository.rb:
```diff
@@ -4,111 +4,111 @@
-  def git_rename_branch(old:, new:)
-    popen_git("branch", "-m", old, new)
+  sig { params(pathname: Pathname, old: String, new: String).void }
+  def self.git_rename_branch(pathname, old:, new:)
```
```suggestion
  def self.rename_branch(pathname, old:, new:)
```

MikeMcQuaid commented on Library/Homebrew/extend/git_repository.rb:
```diff
@@ -4,111 +4,111 @@
-  def git_branch_set_upstream(local:, origin:)
-    popen_git("branch", "-u", "origin/#{origin}", local)
+  sig { params(pathname: Pathname, local: String, origin: String).void }
+  def self.git_branch_set_upstream(pathname, local:, origin:)
```
```suggestion
  def self.set_upstream_branch(pathname, local:, origin:)
```
or
```suggestion
  def self.upstream_branch=(pathname, local:, origin:)
```

MikeMcQuaid commented on Library/Homebrew/extend/git_repository.rb:
```diff
@@ -4,111 +4,111 @@
-  def git_origin_has_branch?(branch)
-    popen_git("ls-remote", "--heads", "origin", branch).present?
+  sig { params(pathname: Pathname, branch: String).returns(T::Boolean) }
+  def self.git_origin_has_branch?(pathname, branch)
```
```suggestion
  def self.origin_has_branch?(pathname, branch)
```

MikeMcQuaid commented on Library/Homebrew/extend/git_repository.rb:
```diff
@@ -4,111 +4,111 @@
-  def git_commit_message(commit = "HEAD", safe: false)
-    popen_git("log", "-1", "--pretty=%B", commit, "--", safe: safe, err: :out)&.strip
+  sig { params(pathname: Pathname, commit: String, safe: T::Boolean).returns(T.nilable(String)) }
+  def self.git_commit_message(pathname, commit = "HEAD", safe: false)
```
```suggestion
  def self.commit_message(pathname, commit = "HEAD", safe: false)
```

MikeMcQuaid commented on Library/Homebrew/extend/git_repository.rb:
```diff
@@ -4,111 +4,111 @@
 # @see Utils::Git
 # @api private
-module GitRepositoryExtension
+module GitRepository
```
```suggestion
module Git
```
or 
```suggestion
module GitPath
```

dduugg replied:
I went with `GitRepoPath` based on feedback from others.

MikeMcQuaid replied:
@dduugg not huge on the `Repo` shortening, personally but I can live with it if you love it.

MikeMcQuaid commented on Library/Homebrew/diagnostic.rb:
```diff
@@ -127,9 +127,9 @@ def please_create_pull_requests(what = "unsupported configuration")
+        return if !Utils::Git.available? || !GitRepository.git?(repository_path)
 
-        current_origin = repository_path.git_origin
+        current_origin = GitRepository.git_origin(repository_path)
```
```suggestion
        current_origin = GitRepository(repository_path).git_origin
```

wondering if something like this would be a nicer API to be able to avoid repeatedly passing the `repository_path` every time and instead create a single instance that you can query with multiple methods.

MikeMcQuaid reviewed (commented):
Nice work @dduugg! Just thinking if you're going to refactor all these call-sites: I think some of the naming/API could be tweaked at the same time to make it a little nicer to use.

dduugg commented:
PTAL @MikeMcQuaid @RandomDSdevel 
(I haven't undertaken any method renaming, since we aren't stamping `Git` in front of every command anymore with this change in implementation)

dduugg commented on Library/Homebrew/extend/git_path.rb:
```diff
@@ -101,14 +101,19 @@ def git_commit_message(commit = "HEAD", safe: false)
   end
 
+  sig { returns(Pathname) }
+  def pathname
```
this is better expressed as `alias pathname __getobj__`, but then we have to add something like [this](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/options.rbi)

RandomDSdevel commented:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Don't mind me, I just had the same thought as the one @MikeMcQuaid did in that review comment of his that I gave a thumbs-up reaction to.  Some other thoughts:  

 - Consider splitting any name changes out of this PR, maybe?  'Git repository' is a sort of domain term of art, so it might be slightly confusing to refer to one as a 'Git path.'  'Git repository's path' or 'path to a/the Git repository' works.  I don't, by any means, necessarily really mean for this to be taken as a blocking comment to any extent, though.  
 - Independently of that, are there any formulas or casks that use this code, by any remote chance?  It could well bw surprising if they did, but that be something to double-check, perhaps?

dduugg commented:
Solid thought, `GitPath` could be confused for the path to the `git` executable itself, so `GitRepoPath` could be less ambiguous choice. The module is marked `@api private`, so I assume it's fair game, but I don't know what kind of escape hatches might exist (Hyrum's Law and all).

carlocab commented on Library/Homebrew/extend/git_path.rb:
```diff
@@ -7,12 +7,14 @@
 # @see Utils::Git
 # @api private
-module GitRepositoryExtension
+class GitPath < SimpleDelegator
```
```suggestion
class GitRepository < SimpleDelegator
```
Maybe? `GitRepoPath` is fine too.

MikeMcQuaid commented:
> Consider splitting any name changes out of this PR, maybe?

I disagree. Makes sense to do all this refactoring at once.

> Independently of that, are there any formulas or casks that use this code, by any remote chance?

Not in homebrew/core. It's not marked as a public API so if it goes 💥 beyond that: that's not on us.

dduugg commented on Library/Homebrew/system_config.rb:
```diff
@@ -69,7 +69,7 @@ def core_tap_branch
     sig { returns(String) }
     def core_tap_origin
-      CoreTap.instance.remote || "(none)"
+      CoreTap.instance.remote
```
`CoreTap.instance.remote` is never falsy (apologies for not adding more comments to this effect on this PR)

MikeMcQuaid replied:
@dduugg No worries, all good! Exactly the type of thing I'm glad for a type system to catch 😅

dduugg commented on Library/Homebrew/extend/git_repo_path.rb:
```diff
@@ -7,12 +7,14 @@
   sig { returns(T::Boolean) }
-  def git?
-    join(".git").exist?
+  def git_repo?
```
`GitRepoPath.new(repo).git_repo?` is repetitive, yes, but the problem is in `Library/Homebrew/tap.rb`. I think some minor redundancy elsewhere is preferable to the below, wdyt?

```diff
diff --git a/Library/Homebrew/tap.rb b/Library/Homebrew/tap.rb
index d1e9c4850..4c17062f6 100644
--- a/Library/Homebrew/tap.rb
+++ b/Library/Homebrew/tap.rb
@@ -135,7 +135,7 @@ class Tap
   def remote
     return default_remote unless installed?

-    @remote ||= path.git_origin
+    @remote ||= path.origin
   end

   # The remote repository name of this {Tap}.
@@ -163,28 +163,28 @@ class Tap

   # True if this {Tap} is a Git repository.
   def git?
-    path.git_repo?
+    path.repo?
   end

   # git branch for this {Tap}.
   def git_branch
     raise TapUnavailableError, name unless installed?

-    path.git_branch
+    path.branch
   end

   # git HEAD for this {Tap}.
   def git_head
     raise TapUnavailableError, name unless installed?

-    @git_head ||= path.git_head
+    @git_head ||= path.head
   end

   # Time since last git commit for this {Tap}.
   def git_last_commit
     raise TapUnavailableError, name unless installed?

-    path.git_last_commit
+    path.last_commit
   end

   # The issues URL of this {Tap}.
@@ -385,20 +385,20 @@ class Tap
       $stderr.ohai "#{name}: changed remote from #{remote} to #{requested_remote}" unless quiet
     end

-    current_upstream_head = path.git_origin_branch
-    return if requested_remote.blank? && path.git_origin_has_branch?(current_upstream_head)
+    current_upstream_head = path.origin_branch
+    return if requested_remote.blank? && path.origin_has_branch?(current_upstream_head)

     args = %w[fetch]
     args << "--quiet" if quiet
     args << "origin"
     safe_system "git", "-C", path, *args
-    path.git_origin_set_head_auto
+    path.origin_set_head_auto

-    new_upstream_head = path.git_origin_branch
+    new_upstream_head = path.origin_branch
     return if new_upstream_head == current_upstream_head

-    path.git_rename_branch old: current_upstream_head, new: new_upstream_head
-    path.git_branch_set_upstream local: new_upstream_head, origin: new_upstream_head
+    path.rename_branch old: current_upstream_head, new: new_upstream_head
+    path.branch_set_upstream local: new_upstream_head, origin: new_upstream_head

     return if quiet
```

MikeMcQuaid replied:
Re-reading the above: yeh, I agree moreso with this that it's confusing that `path` doesn't return a `Pathname`. Feels like this method should have `git_` somewhere in it as it's now returning a `GitRepoPath`

dduugg commented on Library/Homebrew/extend/git_repo_path.rb:
```diff
@@ -7,12 +7,14 @@
+class GitRepoPath < SimpleDelegator
   extend T::Sig
 
+  alias pathname __getobj__
```
Makes sense, done

MikeMcQuaid commented on Library/Homebrew/dev-cmd/extract.rb:
```diff
@@ -113,7 +113,7 @@ def self.extract
     destination_tap.install unless destination_tap.installed?
 
-    repo = source_tap.path
+    repo = source_tap.path.pathname
```
I'm sorry this is taking so much back and forth! 

This now looks pretty weird?

I'd expect that `source_tap.path` returns a `Pathname`. If not, maybe something like

```suggestion
    repo = source_tap.git_repo.path
```

or 

```suggestion
    repo = source_tap.git_repo_path.pathname
```

would make more sense?

dduugg commented:
Going to park this one for a bit in favor of closing out some easier-to-type files

dduugg commented on Library/Homebrew/dev-cmd/pr-pull.rb:
```diff
@@ -156,21 +156,21 @@ def self.determine_bump_subject(old_contents, new_contents, subject_path, reason
 
   # Cherry picks a single commit that modifies a single file.
   # Potentially rewords this commit using {determine_bump_subject}.
-  def self.reword_package_commit(commit, file, reason: "", verbose: false, resolve: false, path: ".")
```
I've renamed `path:` to `git_repo:` and removed the default value (it was never valid bc it was a `String`). Same for `squash_package_commits` below.

dduugg commented:
PTAL @MikeMcQuaid
I finally had the chance to come back to this, and I believe that I've addressed all of the outstanding feedback. Let me know what you think.
`GitRepository` is now a standalone class without delegation. To do this, I added a new ivar to `Tap`, and updated all the call sites depending on whether they were using `git` or ordinary `Pathname` methods. I've updated the method names as well, based on earlier suggestions.

MikeMcQuaid reviewed (approved):
Looks great, thanks for this @dduugg!
